Connects System/Context to SystemBase/ContextBase

sherm1 opened the pull request:
This PR just wires up the existing System framework classes to the new base classes introduced from the caching branch in PRs #8050, #8227, and #8372. Construction and cloning of the Context are the main issues addressed here. There are no functionality changes here and no new unit tests -- there are good tests of construction and cloning in system_test, diagram_test, {leaf,diagram}{_context}_test.

I will be moving a lot of type-agnostic functionality into the SystemBase and ContextBase classes and this PR makes it possible for me to do that without structural changes. The use of default & inherited copy constructors for cloning means I can add members freely to the ContextBase without having to get permission from the derived classes!

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8420)
<!-- Reviewable:end -->

sherm1 commented:
+@jwnimmer-tri for feature review, please.

---

Review status: 0 of 16 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-:-L8cocY_6-pSqyaUfLE4:b-qf68jg)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 18 files at r1, 1 of 1 files at r2, 1 of 7 files at r3, 1 of 3 files at r4.
Review status: 6 of 16 files reviewed at latest revision, all discussions resolved.

---

*[systems/framework/context.h, line 57 at r4](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L8ljvS90xuyoH-ZumaB:-L8ljvS90xuyoH-ZumaC:b-wnua2p) ([raw file](https://github.com/robotlocomotion/drake/blob/b3b8b7c3a37df8794128e3ae3a6cb2ac72c37c28/systems/framework/context.h#L57)):*
> ```objc
>     auto clone_base = ContextBase::Clone();
>     return std::unique_ptr<Context<T>>(
>         dynamic_cast<Context<T>*>(clone_base.release()));
> ```

In theory this should always cast to non-nullptr, because DoCloneWithoutPointers would return the correct type.  But since subclasses of LeafSystem could override that method, its probably worth a fail-fast here that the downcast succeeded.

Also, while not required by our cppguide (yet), I think it would be best to make the resource management here failsafe -- don't release the clone_base until there is definitively a new owner.

Ditto on both for CloneWithoutPointers downcasting below.

---

*[systems/framework/context.h, line 64 at r4](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L8ljlN-6trHmpr4gAYf:-L8ljlN-6trHmpr4gAYg:b-h5qzin) ([raw file](https://github.com/robotlocomotion/drake/blob/b3b8b7c3a37df8794128e3ae3a6cb2ac72c37c28/systems/framework/context.h#L64)):*
> ```objc
>   // This is just an intentional shadowing of the base
>   // class method to return a more convenient type.
>   std::unique_ptr<Context<T>> CloneWithoutPointers() const {
> ```

This method is protected in the base class.  Do we need to elevate it to public here?  If so, perhaps worth a comment to clarify that its intentional vs an oversight.

---

*[systems/framework/context.h, line 409 at r4](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L8lkLZe4dnsJ2UkNzx8:-L8lkLZe4dnsJ2UkNzx9:b-6ys7iw) ([raw file](https://github.com/robotlocomotion/drake/blob/0884b56269095eeeec511666d505c39cd730f8af/systems/framework/context.h#L409)):*
> ```objc
>   /// Returns a deep copy of this Context. The clone's input ports will
>   /// hold deep copies of the data that appears on this context's input ports
>   /// at the time the clone is created.
> ```

This second sentence seems important, but wasn't preserved in the new Clone shadow?

---

*[systems/framework/context.cc, line 9 at r4](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L8loJKLCYtOsdzmmiDy:-L8loJKLCYtOsdzmmiDz:b2t7vkx) ([raw file](https://github.com/robotlocomotion/drake/blob/b3b8b7c3a37df8794128e3ae3a6cb2ac72c37c28/systems/framework/context.cc#L9)):*
> ```cpp
> 
> template <typename T>
> Context<T>::~Context() {}
> ```

While while I love the idea of keeping the dtors isolated in cc files, for a templated class that we intend users to use non-standard T's with, I don't think we should do this.

(Or at least, we shouldn't _only_  do the dtor.  If we want to close the set of scalars, we should move much more code here.  But not in this PR.)

---

*[systems/framework/diagram_context.h, line 118 at r4](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L8loorPCCJHm_VmxEFX:-L8loorPCCJHm_VmxEFY:b-ajacqo) ([raw file](https://github.com/robotlocomotion/drake/blob/b3b8b7c3a37df8794128e3ae3a6cb2ac72c37c28/systems/framework/diagram_context.h#L118)):*
> ```objc
> ///           scalar.
> template <typename T>
> class DiagramContext : public Context<T> {
> ```

BTW The `Context` class overview says that users are forbidden to subclass this.  Use `final` here?

---

*[systems/framework/leaf_context.h, line 109 at r4](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L8lnlfc0WG7NS-7KP7A:-L8lnlfe0NuB-3DbBoTG:b70rita) ([raw file](https://github.com/robotlocomotion/drake/blob/b3b8b7c3a37df8794128e3ae3a6cb2ac72c37c28/systems/framework/leaf_context.h#L109)):*
> ```objc
>   }
> 
>   /// Derived classes should reimplement and replace this; don't invoke it.
> ```

BTW It took me a moment to parse this.  Suggest "don't _recursively_ invoke it" or "don't  _inductively_" invoke it" or something, to make clear we're talking about a subclass override not calling our class's same-named method.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I have some drafts of smaller notes off to the side, but I need a short break, so now seems like a good time to post two big picture note, below.

Each note suggests a direction of rework that (as least tentatively, given that I haven't finished the whole PR yet) seems to require substantial refactoring before this can land.  It might be worth discussing these two points before I try to post the rest of my more detailed notes.  Maybe find me on slack if you want to set up a vcon, after thinking over the questions I posed.

---

Review status: 6 of 16 files reviewed at latest revision, 6 unresolved discussions.

---

*[systems/framework/system_base.h, line 438 at r4](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L8mZncFEUV_5NsLvyho:-L8mZncFEUV_5NsLvyhp:b-vbttcy) ([raw file](https://github.com/robotlocomotion/drake/blob/b3b8b7c3a37df8794128e3ae3a6cb2ac72c37c28/systems/framework/system_base.h#L438)):*
> ```objc
>   // Allocates additional context resources if needed, and validates that the
>   // supplied context is acceptable.
>   void AcquireContextResources(ContextBase* context) const;
> ```

Other than TODOs, from my tracing out it looks like the only thing this is doing is  asserting that the created context is sane (either within the framework, or in delegation to leaf system subclasses).

A sanity check on a _const_ ContextBase seems like a useful idiom to have at hand, but I am skeptical about some "resource acquisition" general strategy give some of the other discussions I've posted (or soon will post) in the rest of this PR.

What do you think about rephrasing this as _only_ a last-look validation check, using a const ContextBase, with downcasting and other Do helpers as needed for subclasses in System<T>, LeafSystem<T>, etc?

I think at that point it could be made protected / private also, which is another bad smell of the current design -- that Acquire is mutating things in a non-idempotent way, but is public and without API docs.

---

*[systems/framework/system_base.cc, line 69 at r4](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L8mV85i4TukYOSeeBPs:-L8mV85i4TukYOSeeBPt:b-tlf42j) ([raw file](https://github.com/robotlocomotion/drake/blob/b3b8b7c3a37df8794128e3ae3a6cb2ac72c37c28/systems/framework/system_base.cc#L69)):*
> ```cpp
> 
>   // We now have a complete Context. We can allocate space for cache entry
>   // values using the allocators, which require a context.
> ```

Having traced and read this all though, I am deeply skeptical that our Allocation functions for both CacheEntry and OutputPort should be passed a context reference.  Given that CacheEntry's live in the Context, and OutputPorts are TODO'd elsewhere in this PR to live in the Context, I do not see how we can straightforwardly specify how to Allocate them, since the context they will receive will by definition be incomplete due to the Catch-22.  I really don't want to try to define "initialization states" that would somehow allow this to happen in phases with semi-valid Contexts being passed around.

Reading though one of the few existing example codes use that does need the context argument during OutputPort allocation, GeometrySystem::MakePoseBundle, it is defectively assuming that the context its given has some specific State value set, and moreover its CalcPoseBundle assumes that the MakePoseBundle result was freshly populated before CalcPoseBundle is called, which is also defective.

In fact that's the only place (on first search) I found where the context argument to port allocation was being used!

Do we have any examples or thoughts on why cache entry allocation should require the context?  I would just as rather punt on it and make everything 10x simpler.

In my mind, Allocate methods are solely about fixing the runtime type for our Value type erasure.  I would be surprised if we had good reasons why that required a Context argument at all -- I suspect that the only possibly use would be to pass MyFooSystem ctor arguments furtively through to the allocation, but really I think lambda capture and/or member fields are most likely correct in that case.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-:-L8may_w9wmOBWAGLbpe:b-gylr9b)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I picked this up again and reviewed a few more ancillary files, but at this point I think basically just the big open questions need a resolution, and then I should be able to finish up the remaining review items relatively quickly.

---

Reviewed 3 of 18 files at r1, 1 of 7 files at r3.
Review status: 10 of 16 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-:-L8n3CRwAv_EcTxZVlOr:b-i8z493)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Thanks -- probably worth f2f discussion but I want to digest your comments first. Preview: context is necessary for allocation of cache (and port) values so that the allocation can depend on port types and parameters at least. For allocations that end up back in the context it would be ideal (conceptually) to supply only those parts of the context that must already be there, but I think that would be unnecessarily complicated and buy nothing of value in actual practice.

---

Review status: 10 of 16 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-:-L8n6GX06w8bcdmFzXYQ:bb9y8ll)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
I do see your point. There is at least a need for a solid contract about what's in the Context when a cache allocator sees it. Likely we can only guarantee that the Context topology is fixed, and that it contains correct sizes and default values for parameters and state variables. Those can in principle be obtained from the Diagram (for topology) and from the local sub-System (for sizes and default values). It's a lot easier to extract that information from the Context because the Diagram has already been mined to build the Context, which is a much simpler data structure. For example, getting size information may require calling a method like System::AllocateTimeDerivatives() and measuring the result. Will see if I can write a compelling contract.

jwnimmer-tri commented:
I am still waiting to hear which specific kind of (or example of) Allocate helper could sensibly depend on a parameter or default value.  In general, in my mental model, Allocation should be about runtime types, not runtime values.

sherm1 commented:
Tidied up a few more things.

---

Review status: 1 of 20 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-:-L9mTLKC8HfMYKvRmnKQ:bo1jxwo)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> pending open discussions.

---

Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[systems/framework/system.h, line 1102 at r8](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9matdN4wGbALBIy6sa:-L9matdN4wGbALBIy6sb:b-gtl8x0) ([raw file](https://github.com/robotlocomotion/drake/blob/308ba36f6daba94c1add0b43bb7a3802aa62a566/systems/framework/system.h#L1102)):*
> ```objc
>       context.VerifyInputPort(this->get_input_port(i));
>     }
>   }
> ```

FYI Not the fault of this PR, but neither of the two checks in this method appear to have any unit tests (i.e., testing that we correctly report such kinds of errors).  Is that worthy of an issue or TODO?

---

*[systems/framework/system_base.h, line 447 at r8](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9m_lCV0CvAfULzvR9d:-L9m_lCV0CvAfULzvR9e:b-p945yp) ([raw file](https://github.com/robotlocomotion/drake/blob/308ba36f6daba94c1add0b43bb7a3802aa62a566/systems/framework/system_base.h#L447)):*
> ```objc
>   }
> ```

BTW Extra blank line?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-:-L9mbFJ1BTh7e3u-rQ3s:bqp17r1)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Addressed all the comments that don't require moving code around so they will be easy to spot.

---

Review status: 14 of 20 files reviewed at latest revision, 13 unresolved discussions.

---

*[systems/framework/diagram.h, line 225 at r7](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mUVhJ2TItrZcO0qPq:-L9mYed-3x7EOA9_prtl:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5c3495d604f6381ca132128a7f6179220cfa9c72/systems/framework/diagram.h#L225)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI While I think this formulation of the cast / demand / etc. is correct in this case (a sufficient set of methods are marked final it might be nice to keep the cast formulation identical to the other instances in this PR, just to forestall stumbles from future readers.
</blockquote></details>

Done.

---

*[systems/framework/diagram.h, line 248 at r7](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mUiqJBOylvHj4Er9f:-L9mZBIW7hNjpdiGhTWB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5c3495d604f6381ca132128a7f6179220cfa9c72/systems/framework/diagram.h#L248)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Don't need `move`.
</blockquote></details>

Done.

---

*[systems/framework/diagram_context.h, line 372 at r7](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mQeXV7mnjlt1zuDFZ:-L9m_gWFBXuAdqU683aA:b-jzfh16) ([raw file](https://github.com/robotlocomotion/drake/blob/5c3495d604f6381ca132128a7f6179220cfa9c72/systems/framework/diagram_context.h#L372)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Don't need `move`.
</blockquote></details>

Done. You're right, but I don't understand why. I see a U-templatized unique_ptr constructor that takes an rvalue but not an lvalue. Can you explain why this works?

---

*[systems/framework/leaf_system.h, line 144 at r8](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mTRpXCNc8N10clos-:-L9maDVi6kXtnw-MvY6-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/308ba36f6daba94c1add0b43bb7a3802aa62a566/systems/framework/leaf_system.h#L144)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Don't need `move`.
</blockquote></details>

Done.

---

*[systems/framework/system.h, line 90 at r7](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mRX2FAX_xCP3Qiqko:-L9mauUV5P4EDfJIXLbs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5c3495d604f6381ca132128a7f6179220cfa9c72/systems/framework/system.h#L90)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI That we are shadowing our _base class's_ method of the same name isn't as obvious in this phrasing as perhaps it could be.
</blockquote></details>

Done.

---

*[systems/framework/system.h, line 96 at r8](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mSnMqA_DfiPR5UNaw:-L9mjEWi1HFjB65sgcCm:bknleqx) ([raw file](https://github.com/robotlocomotion/drake/blob/308ba36f6daba94c1add0b43bb7a3802aa62a566/systems/framework/system.h#L96)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I would love to see this as a helper function in `common` (unique_ptr_cast) to do this for us, though perhaps that's best as a TODO for now, because the edge cases for a generic routine are brutal (StackOverflow has some tips).  Maybe open an issue (you can assign to me)?
</blockquote></details>

Done. Issue #8570. I don't want to sprinkle around TODOs for this now but these will be easy to find if we end up with a satisfactory utility.

---

*[systems/framework/system.h, line 1102 at r8](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9matdN4wGbALBIy6sa:-L9mkuqv40tB2UUhsuP0:btl0a19) ([raw file](https://github.com/robotlocomotion/drake/blob/308ba36f6daba94c1add0b43bb7a3802aa62a566/systems/framework/system.h#L1102)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Not the fault of this PR, but neither of the two checks in this method appear to have any unit tests (i.e., testing that we correctly report such kinds of errors).  Is that worthy of an issue or TODO?
</blockquote></details>

Done. A TODO is enough I think. Next time it gets touched it will have to be tested.

---

*[systems/framework/system_base.h, line 433 at r5](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mO9bk6Y9pLYXVGrdE:-L9mc8ZG8RJXmeD2IaX1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/d164b4ffa48fdd48fe5a1985eed9ef3bffe17f82/systems/framework/system_base.h#L433)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Extra blank line?
</blockquote></details>

Done.

---

*[systems/framework/system_base.h, line 463 at r7](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mSEp51ri_loqSWSA7:-L9meBYc7_how5VT_bD8:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5c3495d604f6381ca132128a7f6179220cfa9c72/systems/framework/system_base.h#L463)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Probably worth some text here contrasting this method vs DoValidateAllocatedContext.
</blockquote></details>

Done.

---

*[systems/framework/system_base.h, line 447 at r8](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9m_lCV0CvAfULzvR9d:-L9mcD7kErTZQSfU2mLt:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/308ba36f6daba94c1add0b43bb7a3802aa62a566/systems/framework/system_base.h#L447)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Extra blank line?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-:-L9mlzAHDhxIfNrMaZhI:bj4byiw)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 6 of 6 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[systems/framework/diagram_context.h, line 372 at r7](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mQeXV7mnjlt1zuDFZ:-L9mr6Iy7o7AE500IPz7:b-uyhk92) ([raw file](https://github.com/robotlocomotion/drake/blob/5c3495d604f6381ca132128a7f6179220cfa9c72/systems/framework/diagram_context.h#L372)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Done. You're right, but I don't understand why. I see a U-templatized unique_ptr constructor that takes an rvalue but not an lvalue. Can you explain why this works?
</blockquote></details>

I think https://stackoverflow.com/questions/14856344/when-should-stdmove-be-used-on-a-function-return-value is on point.  My mental model is "returns by value are always either copy elision (NRVO) or else moves" or in short "never say `return move`".  (Ideally we'd have a clang-tidy linter for this.)

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
All feature review comments addressed, ptal. I left a few curated commits because the last one moves some code around without changing it. Will squash later.
+@seancurtis-tri for platform review please, per rotation.

---

Review status: 15 of 20 files reviewed at latest revision, 3 unresolved discussions.

---

*[systems/framework/diagram.h, line 212 at r7](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mUw434m6OCNu6tTzm:-L9mroAB1fq-l2A3LzFD:b-4pl4ij) ([raw file](https://github.com/robotlocomotion/drake/blob/5c3495d604f6381ca132128a7f6179220cfa9c72/systems/framework/diagram.h#L212)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW This method (and as well below) probably shouldn't be public?
</blockquote></details>

Done. Moved to top of private section.

---

*[systems/framework/diagram_context.h, line 372 at r7](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mQeXV7mnjlt1zuDFZ:-L9mvAgzD2clIfUv8Tss:b-xgn9v4) ([raw file](https://github.com/robotlocomotion/drake/blob/5c3495d604f6381ca132128a7f6179220cfa9c72/systems/framework/diagram_context.h#L372)):*
Wow! This is exactly on point and is a very clever language feature. Definitely going into my collection of obscure C++ trivia! For you lurkers, here's the quote:
> When the criteria for elision of a copy operation are met or would be met save for the fact that the source object is a function parameter, and the object to be copied is designated by an lvalue, overload resolution to select the constructor for the copy is ﬁrst performed as if the object were designated by an rvalue.

---

*[systems/framework/system.h, line 1105 at r8](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mW41o2PUe8NNV-3iq:-L9msHqZ-7SqEotl_UYI:bvlhbre) ([raw file](https://github.com/robotlocomotion/drake/blob/308ba36f6daba94c1add0b43bb7a3802aa62a566/systems/framework/system.h#L1105)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I am not sure this should be `public`?
</blockquote></details>

Done. Quite right -- moved to private.

---

*[systems/framework/system_base.h, line 457 at r5](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9mOOp99884OAAwQ5k8:-L9muWb_-4c61M7zrhi2:bov5j8w) ([raw file](https://github.com/robotlocomotion/drake/blob/d164b4ffa48fdd48fe5a1985eed9ef3bffe17f82/systems/framework/system_base.h#L457)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

~FYI Consider promising as part of SystemBase::DoValidate...'s contract whether or not this method is guaranteed to be a no-op -- in other words, when overriding this _must_ I or _should I not_ invoke this base method.~

~FYI Consider reminding subclasses to call their superclass version of this method (in case there are a cascade of them in a deep hierarchy)?~

_Edit: I guess since Leaf and Diagram both make this final, this isn't a big deal._

Though that makes me wonder, I think this method be abstract?  If Leaf and Diagram both override something with independent implementations, I think we've often been making such base methods abstract, to remove clutter and help clarify precisely what's happening.
</blockquote></details>

Done. Made pure virtual and fixed up the various framework tests that cobble together their own systems that are neither leaf nor diagram.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-:-L9momUQDyRS5lUDI6Hc:b-w3lsm6)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 5 of 5 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Modulo a couple of nit-picks which allow me to earn my keep.

---

Reviewed 1 of 7 files at r3, 10 of 17 files at r5, 1 of 3 files at r8, 3 of 6 files at r9, 5 of 5 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[systems/framework/context.h, line 30 at r10](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9pf5eA8ZyvnQiDfj9Z:-L9pf5eA8ZyvnQiDfj9_:b-zhia21) ([raw file](https://github.com/robotlocomotion/drake/blob/666e0fed64e43482383ff40e0aed613c879c2180/systems/framework/context.h#L30)):*
> ```objc
> 
> /// %Context is an abstract class template that represents all the typed
> /// values that are used in a System's computations: time, and numerical state,
> ```

BTW There seem to be a superfluity of "ands" in this and the next sentence.

---

*[systems/framework/context_base.h, line 146 at r10](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9pRAS3-X6NKZHrI20u:-L9pRAS3-X6NKZHrI20v:bkdtn9t) ([raw file](https://github.com/robotlocomotion/drake/blob/666e0fed64e43482383ff40e0aed613c879c2180/systems/framework/context_base.h#L146)):*
> ```objc
>   ContextBase(const ContextBase&) = default;
> 
>   /** Clones a context but without any of its internal pointers. */
> ```

FYI I'm late to this show. This documentation entered into master in an earlier PR. But it caught my attention so I'll make a passing remark. I don't care much for this comment because "without any of its internal pointers" seems inaccurate/incomplete. The clone context *has* pointers -- but their *values* are null. Right? I.e., the full structure (number of pointers, etc.) are cloned, but all pointer values are null. If my understanding is correct, I don't feel the comment aligns with that understanding.

---

*[systems/framework/leaf_system.h, line 96 at r10](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9po2FFBuk1fFuLMDhi:-L9po2FFBuk1fFuLMDhj:bm9gnlu) ([raw file](https://github.com/robotlocomotion/drake/blob/666e0fed64e43482383ff40e0aed613c879c2180/systems/framework/leaf_system.h#L96)):*
> ```objc
>   }
> 
>   /// Shadows `System<T>::AllocateContext` to provide a more concrete return
> ```

BTW See other comment on backticks.

---

*[systems/framework/system.h, line 90 at r10](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9pgloC8jdc7a4BoV0O:-L9pgloD3h3ODrs99Ho2:bjbc5t4) ([raw file](https://github.com/robotlocomotion/drake/blob/666e0fed64e43482383ff40e0aed613c879c2180/systems/framework/system.h#L90)):*
> ```objc
>   //@{
> 
>   /// Returns a `Context<T>` suitable for use with this `System<T>`.
> ```

FYI I'm unsure why there are backticks. In the rendered doxygen, the links are still valid, but, in this case, the class names are uniquely rendered with fixed-width font.

---

*[systems/framework/vector_system.h, line 281 at r10](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-L9pnHR1EwHN4h5eObwa:-L9pnHR1EwHN4h5eObwb:b-6mf1mo) ([raw file](https://github.com/robotlocomotion/drake/blob/666e0fed64e43482383ff40e0aed613c879c2180/systems/framework/vector_system.h#L281)):*
> ```objc
>     DRAKE_DEMAND(context.get_num_input_ports() <= 1);
> 
>     // At most one of either continuous xor discrete state.
> ```

FYI The language "At most one of either" renders the technical "xor" a bit moot. The sentence loses no clarity (and probably gains) in swapping "xor" for "or".

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-:-L9pou46D3PN12H3szrQ:buxkfr5)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Silly reviewable.

---

Reviewed 5 of 5 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/robotlocomotion/drake/8420#-:-L9qPH9kEKhTQRHtVen1:b-ma1j0m)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Thanks, @jwnimmer-tri and @SeanCurtis-TRI. Last change was just to comments so merging now.
